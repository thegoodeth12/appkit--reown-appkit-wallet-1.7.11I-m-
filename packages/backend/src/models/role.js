import { ValidationError } from 'objection';
import Base from '@/models/base.js';
import Permission from '@/models/permission.js';
import User from '@/models/user.js';
import SamlAuthProvider from '@/models/saml-auth-provider.ee.js';
import NotAuthorizedError from '@/errors/not-authorized.js';

class Role extends Base {
  static tableName = 'roles';

  static jsonSchema = {
    type: 'object',
    required: ['name'],

    properties: {
      id: { type: 'string', format: 'uuid' },
      name: { type: 'string', minLength: 1 },
      description: { type: ['string', 'null'], maxLength: 255 },
      createdAt: { type: 'string' },
      updatedAt: { type: 'string' },
    },
  };

  static relationMappings = () => ({
    users: {
      relation: Base.HasManyRelation,
      modelClass: User,
      join: {
        from: 'roles.id',
        to: 'users.role_id',
      },
    },
    permissions: {
      relation: Base.HasManyRelation,
      modelClass: Permission,
      join: {
        from: 'roles.id',
        to: 'permissions.role_id',
      },
    },
  });

  static get virtualAttributes() {
    return ['isAdmin'];
  }

  get isAdmin() {
    return this.name === 'Admin';
  }

  static async findAdmin() {
    return await this.query().findOne({ name: 'Admin' });
  }

  async preventAlteringAdmin() {
    const currentRole = await Role.query().findById(this.id);

    if (currentRole.isAdmin) {
      throw new NotAuthorizedError('The admin role cannot be altered!');
    }
  }

  async deletePermissions() {
    return await this.$relatedQuery('permissions').delete();
  }

  async createPermissions(permissions) {
    if (permissions?.length) {
      const validPermissions = Permission.filter(permissions).map(
        (permission) => ({
          ...permission,
          roleId: this.id,
        })
      );

      await Permission.query().insert(validPermissions);
    }
  }

  async updatePermissions(permissions) {
    await this.deletePermissions();

    await this.createPermissions(permissions);
  }

  async updateWithPermissions(data) {
    const { name, description, permissions } = data;

    await this.updatePermissions(permissions);

    await this.$query().patchAndFetch({
      id: this.id,
      name,
      description,
    });

    return await this.$query()
      .leftJoinRelated({
        permissions: true,
      })
      .withGraphFetched({
        permissions: true,
      });
  }

  async deleteWithPermissions() {
    await this.deletePermissions();

    return await this.$query().delete();
  }

  async assertNoRoleUserExists() {
    const userCount = await this.$relatedQuery('users').limit(1).resultSize();
    const hasUsers = userCount > 0;

    if (hasUsers) {
      throw new ValidationError({
        data: {
          role: [
            {
              message: `All users must be migrated away from the "${this.name}" role.`,
            },
          ],
        },
        type: 'ValidationError',
      });
    }
  }

  async assertNoConfigurationUsage() {
    const samlAuthProviderUsingDefaultRole = await SamlAuthProvider.query()
      .where({
        default_role_id: this.id,
      })
      .limit(1)
      .first();

    if (samlAuthProviderUsingDefaultRole) {
      throw new ValidationError({
        data: {
          samlAuthProvider: [
            {
              message:
                'You need to change the default role in the SAML configuration before deleting this role.',
            },
          ],
        },
        type: 'ValidationError',
      });
    }
  }

  async assertRoleIsNotUsed() {
    await this.assertNoRoleUserExists();

    await this.assertNoConfigurationUsage();
  }

  async $beforeUpdate(opt, queryContext) {
    await super.$beforeUpdate(opt, queryContext);

    await this.preventAlteringAdmin();
  }

  async $beforeDelete(queryContext) {
    await super.$beforeDelete(queryContext);

    await this.preventAlteringAdmin();

    await this.assertRoleIsNotUsed();
  }
}

export default Role;
