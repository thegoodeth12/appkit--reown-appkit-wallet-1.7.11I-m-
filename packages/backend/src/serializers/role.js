import permissionSerializer from '@/serializers/permission.js';

const roleSerializer = (role) => {
  let roleData = {
    id: role.id,
    name: role.name,
    key: role.key,
    description: role.description,
    createdAt: role.createdAt.getTime(),
    updatedAt: role.updatedAt.getTime(),
    isAdmin: role.isAdmin,
  };

  if (role.permissions?.length > 0) {
    roleData.permissions = role.permissions.map((permission) =>
      permissionSerializer(permission)
    );
  }

  return roleData;
};

export default roleSerializer;
