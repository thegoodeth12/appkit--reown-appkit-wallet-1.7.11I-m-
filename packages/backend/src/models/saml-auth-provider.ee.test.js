import { vi, beforeEach, describe, it, expect } from 'vitest';
import { v4 as uuidv4 } from 'uuid';
import SamlAuthProvider from '@/models/saml-auth-provider.ee.js';
import RoleMapping from '@/models/role-mapping.ee.js';
import axios from '@/helpers/axios-with-proxy.js';
import Identity from '@/models/identity.ee.js';
import Base from '@/models/base.js';
import appConfig from '@/config/app.js';
import { createSamlAuthProvider } from '@/factories/saml-auth-provider.ee.js';
import { createRoleMapping } from '@/factories/role-mapping.js';
import { createRole } from '@/factories/role.js';

describe('SamlAuthProvider model', () => {
  it('tableName should return correct name', () => {
    expect(SamlAuthProvider.tableName).toBe('saml_auth_providers');
  });

  it('jsonSchema should have the correct schema', () => {
    expect(SamlAuthProvider.jsonSchema).toMatchSnapshot();
  });

  it('relationMappings should return correct associations', () => {
    const relationMappings = SamlAuthProvider.relationMappings();

    const expectedRelations = {
      identities: {
        relation: Base.HasOneRelation,
        modelClass: Identity,
        join: {
          from: 'identities.provider_id',
          to: 'saml_auth_providers.id',
        },
      },
      roleMappings: {
        relation: Base.HasManyRelation,
        modelClass: RoleMapping,
        join: {
          from: 'saml_auth_providers.id',
          to: 'role_mappings.saml_auth_provider_id',
        },
      },
    };

    expect(relationMappings).toStrictEqual(expectedRelations);
  });

  it('virtualAttributes should return correct attributes', () => {
    const virtualAttributes = SamlAuthProvider.virtualAttributes;

    const expectedAttributes = ['loginUrl', 'remoteLogoutUrl'];

    expect(virtualAttributes).toStrictEqual(expectedAttributes);
  });

  it('loginUrl should return the URL of login', () => {
    const samlAuthProvider = new SamlAuthProvider();
    samlAuthProvider.issuer = 'sample-issuer';

    vi.spyOn(appConfig, 'baseUrl', 'get').mockReturnValue(
      'https://automatisch.io'
    );

    expect(samlAuthProvider.loginUrl).toStrictEqual(
      'https://automatisch.io/login/saml/sample-issuer'
    );
  });

  it('loginCallbackUrl should return the URL of login callback', () => {
    const samlAuthProvider = new SamlAuthProvider();
    samlAuthProvider.issuer = 'sample-issuer';

    vi.spyOn(appConfig, 'baseUrl', 'get').mockReturnValue(
      'https://automatisch.io'
    );

    expect(samlAuthProvider.loginCallBackUrl).toStrictEqual(
      'https://automatisch.io/login/saml/sample-issuer/callback'
    );
  });

  it('remoteLogoutUrl should return the URL from entrypoint', () => {
    const samlAuthProvider = new SamlAuthProvider();
    samlAuthProvider.entryPoint = 'https://example.com/saml/logout';

    expect(samlAuthProvider.remoteLogoutUrl).toStrictEqual(
      'https://example.com/saml/logout'
    );
  });

  it('config should return the correct configuration object', () => {
    const samlAuthProvider = new SamlAuthProvider();

    samlAuthProvider.certificate = 'sample-certificate';
    samlAuthProvider.signatureAlgorithm = 'sha256';
    samlAuthProvider.entryPoint = 'https://example.com/saml';
    samlAuthProvider.issuer = 'sample-issuer';

    vi.spyOn(appConfig, 'baseUrl', 'get').mockReturnValue(
      'https://automatisch.io'
    );

    const expectedConfig = {
      callbackUrl: 'https://automatisch.io/login/saml/sample-issuer/callback',
      cert: 'sample-certificate',
      entryPoint: 'https://example.com/saml',
      issuer: 'sample-issuer',
      signatureAlgorithm: 'sha256',
      logoutUrl: 'https://example.com/saml',
    };

    expect(samlAuthProvider.config).toStrictEqual(expectedConfig);
  });

  it('generateLogoutRequestBody should return a correctly encoded SAML logout request', () => {
    vi.mock('uuid', () => ({
      v4: vi.fn(),
    }));

    const samlAuthProvider = new SamlAuthProvider();

    samlAuthProvider.entryPoint = 'https://example.com/saml';
    samlAuthProvider.issuer = 'sample-issuer';

    const mockUuid = '123e4567-e89b-12d3-a456-426614174000';
    uuidv4.mockReturnValue(mockUuid);

    const sessionId = 'test-session-id';

    const logoutRequest = samlAuthProvider.generateLogoutRequestBody(sessionId);

    const expectedLogoutRequest = `
      <samlp:LogoutRequest
          xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
          ID="${mockUuid}"
          Version="2.0"
          IssueInstant="${new Date().toISOString()}"
          Destination="https://example.com/saml">

          <saml:Issuer xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">sample-issuer</saml:Issuer>
          <samlp:SessionIndex>test-session-id</samlp:SessionIndex>
      </samlp:LogoutRequest>
    `;

    const expectedEncodedRequest = Buffer.from(expectedLogoutRequest).toString(
      'base64'
    );

    expect(logoutRequest).toBe(expectedEncodedRequest);
  });

  it('terminateRemoteSession should send the correct POST request and return the response', async () => {
    vi.mock('../helpers/axios-with-proxy.js', () => ({
      default: {
        post: vi.fn(),
      },
    }));

    const samlAuthProvider = new SamlAuthProvider();

    samlAuthProvider.entryPoint = 'https://example.com/saml';
    samlAuthProvider.generateLogoutRequestBody = vi
      .fn()
      .mockReturnValue('mockEncodedLogoutRequest');

    const sessionId = 'test-session-id';

    const mockResponse = { data: 'Logout Successful' };
    axios.post.mockResolvedValue(mockResponse);

    const response = await samlAuthProvider.terminateRemoteSession(sessionId);

    expect(samlAuthProvider.generateLogoutRequestBody).toHaveBeenCalledWith(
      sessionId
    );

    expect(axios.post).toHaveBeenCalledWith(
      'https://example.com/saml',
      'SAMLRequest=mockEncodedLogoutRequest',
      {
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded',
        },
      }
    );

    expect(response).toBe(mockResponse);
  });

  describe('updateRoleMappings', () => {
    let samlAuthProvider;

    beforeEach(async () => {
      samlAuthProvider = await createSamlAuthProvider();
    });

    it('should remove all existing role mappings', async () => {
      await createRoleMapping({
        samlAuthProviderId: samlAuthProvider.id,
        remoteRoleName: 'Admin',
      });

      await createRoleMapping({
        samlAuthProviderId: samlAuthProvider.id,
        remoteRoleName: 'User',
      });

      await samlAuthProvider.updateRoleMappings([]);

      const roleMappings = await samlAuthProvider.$relatedQuery('roleMappings');
      expect(roleMappings).toStrictEqual([]);
    });

    it('should return the updated role mappings when new ones are provided', async () => {
      const adminRole = await createRole({ name: 'Admin' });
      const userRole = await createRole({ name: 'User' });

      const newRoleMappings = [
        { remoteRoleName: 'Admin', roleId: adminRole.id },
        { remoteRoleName: 'User', roleId: userRole.id },
      ];

      const result = await samlAuthProvider.updateRoleMappings(newRoleMappings);

      const refetchedRoleMappings = await samlAuthProvider.$relatedQuery(
        'roleMappings'
      );

      expect(result).toStrictEqual(refetchedRoleMappings);
    });
  });
});
