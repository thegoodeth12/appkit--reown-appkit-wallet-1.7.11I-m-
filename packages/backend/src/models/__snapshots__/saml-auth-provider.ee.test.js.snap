// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SamlAuthProvider model > jsonSchema should have the correct schema 1`] = `
{
  "properties": {
    "active": {
      "type": "boolean",
    },
    "certificate": {
      "minLength": 1,
      "type": "string",
    },
    "defaultRoleId": {
      "format": "uuid",
      "type": "string",
    },
    "emailAttributeName": {
      "minLength": 1,
      "type": "string",
    },
    "entryPoint": {
      "minLength": 1,
      "type": "string",
    },
    "firstnameAttributeName": {
      "minLength": 1,
      "type": "string",
    },
    "id": {
      "format": "uuid",
      "type": "string",
    },
    "issuer": {
      "minLength": 1,
      "type": "string",
    },
    "name": {
      "minLength": 1,
      "type": "string",
    },
    "roleAttributeName": {
      "minLength": 1,
      "type": "string",
    },
    "signatureAlgorithm": {
      "enum": [
        "sha1",
        "sha256",
        "sha512",
      ],
      "type": "string",
    },
    "surnameAttributeName": {
      "minLength": 1,
      "type": "string",
    },
  },
  "required": [
    "name",
    "certificate",
    "signatureAlgorithm",
    "entryPoint",
    "issuer",
    "firstnameAttributeName",
    "surnameAttributeName",
    "emailAttributeName",
    "roleAttributeName",
    "defaultRoleId",
  ],
  "type": "object",
}
`;
